src/cliffordsynthesis/encoding/TwoQubitEncoder.cpp

```diff
@@ -0,0 +1,251 @@
+//
+// Created by Velsh Aleksei on 16.06.23.
+//
+
+#include "cliffordsynthesis/encoding/TwoQubitEncoder.hpp"
+
+#include "LogicTerm/LogicTerm.hpp"
+#include "operations/OpType.hpp"
+#include "utils/logging.hpp"
+
+namespace cs::encoding {
+
+using namespace logicbase;
+
+void encoding::TwoQubitEncoder::collectTwoQubitGateVariables(
+    const std::size_t pos, const std::size_t qubit, const bool target,
+    LogicVector& variables) const {
+  const auto& twoQubitGates = vars.gC[pos];
+  const auto  n             = twoQubitGates.size();
+  for (std::size_t q = 0; q < n; ++q) {
+    if (q == qubit) {
+      if (!target) {
+        variables.emplace_back(twoQubitGates[qubit][qubit]);
+      }
+      continue;
+    }
+    if (target) {
+      variables.emplace_back(twoQubitGates[q][qubit]);
+    } else {
+      variables.emplace_back(twoQubitGates[qubit][q]);
+    }
+  }
+}
+
+void encoding::TwoQubitEncoder::assertConsistency() const {
+  DEBUG() << "Asserting gate consistency";
+  for (std::size_t t = 0U; t <= T; ++t) {
+    // asserting only a single gate is applied on each qubit.
+    for (std::size_t q = 0U; q < N; ++q) {
+      LogicVector singleQubitGateVariables{};
+      LogicVector twoQubitGateVariables{};
+      LogicVector pauliGateVariables{};
+      if (t == T) {
+        vars.collectPauliQubitGateVariables(q, pauliGateVariables);
+        assertExactlyOne(pauliGateVariables);
+        IF_PLOG(plog::verbose) {
+          TRACE() << "Pauli Qubit Gate variables at time " << t << " and qubit "
+                  << q;
+          for (const auto& var : pauliGateVariables) {
+            TRACE() << var.getName();
+          }
+        }
+      } else if (t % 2 == 0) {
+        vars.collectSingleQubitGateVariables(t / 2, q,
+                                             singleQubitGateVariables);
+        assertExactlyOne(singleQubitGateVariables);
+        IF_PLOG(plog::verbose) {
+          TRACE() << "Single Qubit Gate variables at time " << t
+                  << " and qubit " << q;
+          for (const auto& var : singleQubitGateVariables) {
+            TRACE() << var.getName();
+          }
+        }
+      } else {
+        collectTwoQubitGateVariables(t / 2, q, true, twoQubitGateVariables);
+        collectTwoQubitGateVariables(t / 2, q, false, twoQubitGateVariables);
+        assertExactlyOne(twoQubitGateVariables);
+        IF_PLOG(plog::verbose) {
+          std::cout << twoQubitGateVariables.size() << std::endl;
+
+          TRACE() << "Two Qubit Gate variables at time " << t << " and qubit "
+                  << q;
+          for (const auto& var : twoQubitGateVariables) {
+            TRACE() << var.getName();
+          }
+        }
+      }
+    }
+  }
+}
+
+void encoding::TwoQubitEncoder::assertGateConstraints() {
```

Comment:
After playing around with the code a bit I am now 100% convinced that a last single-qubit gate layer is need after the last two-qubit gate layer. Otherwise depth-optimality cannot be guaranteed.